src/resourcemanager/resourcemanager.cpp

```diff
@@ -16,21 +16,282 @@ namespace aos::zephyr::resourcemanager {
+    size_t      devicesLen;
+    Resource    resources[aos::cMaxNumNodeResources];
+    size_t      resourcesLen;
+    const char* labels[aos::cMaxNumNodeLabels];
+    size_t      labelsLen;
     uint32_t    priority = 0;
 };
 
+} // namespace
+
+// Device.
+
+static const struct json_obj_descr cDeviceDescr[] = {
+    JSON_OBJ_DESCR_PRIM(Device, name, JSON_TOK_STRING),
+    JSON_OBJ_DESCR_ARRAY(Device, hostDevices, aos::cMaxNumHostDevices, hostDevicesLen, JSON_TOK_STRING),
+    JSON_OBJ_DESCR_PRIM(Device, sharedCount, JSON_TOK_NUMBER),
+    JSON_OBJ_DESCR_ARRAY(Device, groups, aos::cMaxNumGroups, groupsLen, JSON_TOK_STRING),
+};
+
+// FS mount.
+
+static const struct json_obj_descr cFSMountDescr[] = {
+    JSON_OBJ_DESCR_PRIM(FSMount, destination, JSON_TOK_STRING),
+    JSON_OBJ_DESCR_PRIM(FSMount, source, JSON_TOK_STRING),
+    JSON_OBJ_DESCR_PRIM(FSMount, type, JSON_TOK_STRING),
+    JSON_OBJ_DESCR_ARRAY(FSMount, options, aos::cFSMountMaxNumOptions, optionsLen, JSON_TOK_STRING),
+};
+
+// Host.
+
+static const struct json_obj_descr cHostDescr[] = {
+    JSON_OBJ_DESCR_PRIM(Host, ip, JSON_TOK_STRING),
+    JSON_OBJ_DESCR_PRIM(Host, hostName, JSON_TOK_STRING),
+};
+
+// Resource.
+
+static const struct json_obj_descr cResourceDescr[] = {
+    JSON_OBJ_DESCR_PRIM(Resource, name, JSON_TOK_STRING),
+    JSON_OBJ_DESCR_ARRAY(Resource, groups, aos::cMaxNumGroups, groupsLen, JSON_TOK_STRING),
+    JSON_OBJ_DESCR_OBJ_ARRAY(
+        Resource, mounts, aos::cMaxNumFSMounts, mountsLen, cFSMountDescr, ARRAY_SIZE(cFSMountDescr)),
+    JSON_OBJ_DESCR_ARRAY(Resource, env, aos::cMaxNumEnvVariables, envLen, JSON_TOK_STRING),
+    JSON_OBJ_DESCR_OBJ_ARRAY(Resource, hosts, aos::cMaxNumHosts, hostsLen, cHostDescr, ARRAY_SIZE(cHostDescr)),
+};
+
+// Node config.
+
 static const struct json_obj_descr cNodeConfigDescr[] = {
     JSON_OBJ_DESCR_PRIM(NodeConfig, version, JSON_TOK_STRING),
     JSON_OBJ_DESCR_PRIM(NodeConfig, nodeType, JSON_TOK_STRING),
+    JSON_OBJ_DESCR_OBJ_ARRAY(
+        NodeConfig, devices, aos::cMaxNumDevices, devicesLen, cDeviceDescr, ARRAY_SIZE(cDeviceDescr)),
+    JSON_OBJ_DESCR_OBJ_ARRAY(
+        NodeConfig, resources, aos::cMaxNumNodeResources, resourcesLen, cResourceDescr, ARRAY_SIZE(cResourceDescr)),
+    JSON_OBJ_DESCR_ARRAY(NodeConfig, labels, aos::cMaxNumNodeLabels, labelsLen, JSON_TOK_STRING),
     JSON_OBJ_DESCR_PRIM(NodeConfig, priority, JSON_TOK_NUMBER),
 };
 
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+static void FillJSONStruct(const Array<DeviceInfo>& inDevices, NodeConfig& jsonNodeConfig)
+{
+    jsonNodeConfig.devicesLen = inDevices.Size();
+
+    for (size_t i = 0; i < inDevices.Size(); ++i) {
+        const auto& inDevice   = inDevices[i];
+        auto&       jsonDevice = jsonNodeConfig.devices[i];
```

Comment:
should we check is `inDevices.Size() <= len jsonNodeConfig.devices` ?

Reply:
Didn't get the point.
Do you mean array max size checks? (for json decode into aos it's done via Array::Resize calls, FillAosStruct for example).

For json encode from aos, no check are done. But anyway our structs rely on aos consts, as well as JSON object desriptors that are declared static in this cpp.

BTW, ocispec doesn't check boundaries when aos is converted into JSON as well